packages/vertexai/src/types/error.ts

```diff
@@ -0,0 +1,110 @@
+import { FirebaseError } from "@firebase/util";
+
+/**
+ * Details object that may be included in an error response.
+ *
+ * @public
+ */
+export interface ErrorDetails {
+  '@type'?: string;
+
+  /** The reason for the error. */
+  reason?: string;
+
+  /** The domain where the error occured. */
+  domain?: string;
+
+  /** Additonal metadata about the error. */
+  metadata?: Record<string, unknown>;
+
+  /** Any other relevant information about the error. */
+  [key: string]: unknown;
+}
+
+/**
+ * Standardized error codes that {@link VertexAIError} can have.
+ *
+ * @public
+ */
+export const enum VertexAIErrorCode {
+  /** A generic error occured. */
+  ERROR = 'error',
+
+  /** An error occurred in a request. */
+  REQUEST_ERROR = 'request-error',
+
+  /** An error occured in a response. */
+  RESPONSE_ERROR = 'response-error',
+
+  /** An error occurred while performing a fetch. */
+  FETCH_ERROR = 'fetch-error',
+
+  /** An error associated with a Content object.  */
+  INVALID_CONTENT = 'invalid-content',
+
+  /** An error occured due to a missing api key. */
+  NO_API_KEY = 'no-api-key',
+
+  /** An error occurred due to a missing model. */
+  NO_MODEL = 'no-model',
+
+  /** An error occured due to a missing project id. */
+  NO_PROJECT_ID = 'no-project-id',
+
+  /** An error occured while parsing. */
+  PARSE_FAILED = 'parse-failed'
+}
+
+
+/**
+ * Error class for the Firebase VertexAI SDK.
+ *
+ * @public
+ */
+export class VertexAIError extends FirebaseError {
+  /**
+   * Stack trace of the error.
+   */
+  readonly stack?: string;
+
+  /**
+   * Creates a new VertexAIError instance.
+   *
+   * @param code - The error code from {@link VertexAIErrorCode}.
+   * @param message - A human-readable message describing the error.
+   * @param status - Optional HTTP status code of the error response.
+   * @param statusText - Optional HTTP status text of the error response.
+   * @param errorDetails - Optional additional details about the error.
+   */
+  constructor(
+    readonly code: VertexAIErrorCode,
```

Comment:
Realized the constructor is already kind of long and only going to get longer as we think of more properties to add. (We already have to add "response" as mentioned in another comment.) Maybe we can put all the optional properties into an object, so it's `constructor(code, message, customErrorData)` where `customErrorData` is an object with all the optional properties.

Reply:
Agreed! since `status`, `statusText`, and `errorDetails` are used only when throwing on HTTP responses, I moved them into a `HTTPErrorDetails` interface, and accept that as a parameter instead.

New constructor:
```javascript
constructor(
    readonly code: VertexAIErrorCode,
    readonly message: string,
    readonly httpErrorDetails?: HTTPErrorDetails,
    readonly generateContentResponse?: GenerateContentResponse
  )
```

New error declaration with http details:
```javascript
      throw new VertexAIError(
        VertexAIErrorCode.FETCH_ERROR,
        `Error fetching from ${url}: [${response.status} ${response.statusText}] ${message}`,
        {
          status: response.status,
          statusText: response.statusText,
          errorDetails
        }
      );
```

Reply:
I think we should future proof it by making it not HTTPError specific. We might put other details in here in the future not related to HTTP, I don't think we should then create a separate arg for that. I think response should go in here too.

Reply:
You don't have to use undefined placeholders if it's an object right? Only if it's function args.

Reply:
Oh, nevermind then!